diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7ee3169..fd47a42 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -9,7 +9,9 @@ cmake_policy(SET CMP0148 NEW) # use modern Python discovery.
 include(GNUInstallDirs)
 
 # set the C++ standard for the project.
-set(CMAKE_CXX_STANDARD 20)
+set(CMAKE_CXX_STANDARD 23)
+# Lix headers forbid NDEBUG
+add_compile_options(-UNDEBUG)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 # dependency discovery
@@ -41,7 +43,7 @@ add_executable(xlix
 )
 
 # explicitly state the C++ standard requirement for the target.
-target_compile_features(xlix PRIVATE cxx_std_20)
+target_compile_features(xlix PRIVATE cxx_std_23)
 
 # add all required include directories for the kernel.
 target_include_directories(xlix PRIVATE
diff --git a/src/lix_eval_helpers.cpp b/src/lix_eval_helpers.cpp
index d16cb98..b539043 100644
--- a/src/lix_eval_helpers.cpp
+++ b/src/lix_eval_helpers.cpp
@@ -113,7 +113,7 @@ namespace xeus_lix
         }
         // parse and evaluate the expression within the current environment
         auto& expr = m_evaluator->parseExprFromString(std::string(expr_str), nix::CanonPath::fromCwd(), m_staticEnv);
-        expr.eval(*m_evalState, *m_localEnv, result);
+        result = expr.eval(*m_evalState, *m_localEnv);
         // force the result to a weak head normal form
         m_evalState->forceValue(result, nix::noPos);
     }
@@ -139,9 +139,9 @@ namespace xeus_lix
             return markdown_stream.str();
         }
         // try to get the documentation comment if it's a lambda with a source position
-        else if (v.isLambda() && v.lambda.fun->pos != nix::noPos)
+        else if (v.isLambda() && v.lambda().fun->pos != nix::noPos)
         {
-            auto pos = m_evaluator->positions[v.lambda.fun->pos];
+            auto pos = m_evaluator->positions[v.lambda().fun->pos];
             if (auto path = std::get_if<nix::CheckedSourcePath>(&pos.origin))
             {
                 auto docComment = lambdaDocsForPos(*path, pos);
diff --git a/src/lix_interpreter.cpp b/src/lix_interpreter.cpp
index 3e68111..fd97555 100644
--- a/src/lix_interpreter.cpp
+++ b/src/lix_interpreter.cpp
@@ -96,16 +96,14 @@ namespace xeus_lix
                     [&](nix::ExprReplBindings& bindings) {
                         for (auto& [name, expr] : bindings.symbols)
                         {
-                            nix::Value* val = m_evaluator->mem.allocValue();
-                            expr->eval(*m_evalState, *m_localEnv, *val);
+                            m_localEnv->values[m_displacement] = expr->eval(*m_evalState, *m_localEnv);
                             (void)expr.release();
                             m_staticEnv->vars.insert_or_assign(name, m_displacement);
-                            m_localEnv->values[m_displacement++] = val;
+                            m_displacement++;
                         }
                     },
                     [&](std::unique_ptr<nix::Expr>& expr) {
-                        nix::Value val(nix::Value::null_t{});
-                        expr->eval(*m_evalState, *m_localEnv, val);
+                        nix::Value val = expr->eval(*m_evalState, *m_localEnv);
                         (void)expr.release();
 
                         nl::json pub_data;
@@ -118,18 +116,17 @@ namespace xeus_lix
                             if (val.type() == nix::nAttrs)
                             {
                                 // a `_toMime` attribute signals a rich representation
-                                if (auto it = val.attrs->find(m_evaluator->symbols.create("_toMime"));
-                                    it != val.attrs->end())
+                                if (auto attr = val.attrs()->get(m_evaluator->symbols.create("_toMime")))
                                 {
-                                    nix::Value& mime_set_val = *it->value;
-                                    m_evalState->forceAttrs(mime_set_val, it->pos, "while rendering _toMime");
+                                    nix::Value& mime_set_val = attr->value;
+                                    m_evalState->forceAttrs(mime_set_val, attr->pos, "while rendering _toMime");
 
                                     nl::json data_bundle;
-                                    for (const auto& attr : *mime_set_val.attrs)
+                                    for (const auto& attr : *mime_set_val.attrs())
                                     {
                                         std::string_view mime_type_sv = m_evaluator->symbols[attr.name];
                                         std::string mime_type(mime_type_sv);
-                                        nix::Value& data_val_ref = *attr.value;
+                                        nix::Value& data_val_ref = attr.value;
                                         std::string data_str(
                                             m_evalState->forceString(data_val_ref, attr.pos, "mime data")
                                         );
@@ -400,7 +397,7 @@ namespace xeus_lix
 
                 if (base_val.type() == nix::nAttrs)
                 {
-                    for (const auto& attr : *base_val.attrs)
+                    for (const auto& attr : *base_val.attrs())
                     {
                         std::string_view attr_name_sv = m_evaluator->symbols[attr.name];
                         std::string attr_name(attr_name_sv);
diff --git a/src/lix_logger.cpp b/src/lix_logger.cpp
index 70e58c8..cdeba30 100644
--- a/src/lix_logger.cpp
+++ b/src/lix_logger.cpp
@@ -8,7 +8,7 @@ namespace xeus_lix
     }
 
     // called by lix to log general messages.
-    void JupyterLogger::log(nix::Verbosity lvl, const std::string_view s)
+    nix::Logger::BufferState JupyterLogger::log(nix::Verbosity lvl, const std::string_view s)
     {
         // error/warn: redirected to stderr of cell
         if (lvl <= nix::lvlWarn) {
@@ -19,21 +19,23 @@ namespace xeus_lix
             p_interpreter->publish_stream("stdout", std::string(s) + "\n");
         }
         // talkative+: ignore
+        return nix::Logger::BufferState::HasSpace;
     }
 
     // called by lix to log structured error information.
-    void JupyterLogger::logEI(const nix::ErrorInfo& ei)
+    nix::Logger::BufferState JupyterLogger::logEI(const nix::ErrorInfo& ei)
     {
         std::stringstream oss;
         // use lix's own error formatting utility.
         nix::showErrorInfo(oss, ei, nix::loggerSettings.showTrace.get());
         // ErrorInfo structures are typically more severe and should always be shown.
         p_interpreter->publish_stream("stderr", oss.str());
+        return nix::Logger::BufferState::HasSpace;
     }
 
     // we don't need the following methods for now
 
-    void JupyterLogger::startActivity(
+    nix::Logger::BufferState JupyterLogger::startActivityImpl(
         nix::ActivityId,
         nix::Verbosity,
         nix::ActivityType,
@@ -42,13 +44,16 @@ namespace xeus_lix
         nix::ActivityId
     )
     {
+        return nix::Logger::BufferState::HasSpace;
     }
 
-    void JupyterLogger::stopActivity(nix::ActivityId)
+    nix::Logger::BufferState JupyterLogger::stopActivityImpl(nix::ActivityId)
     {
+        return nix::Logger::BufferState::HasSpace;
     }
 
-    void JupyterLogger::result(nix::ActivityId, nix::ResultType, const Fields&)
+    nix::Logger::BufferState JupyterLogger::resultImpl(nix::ActivityId, nix::ResultType, const Fields&)
     {
+        return nix::Logger::BufferState::HasSpace;
     }
 }
diff --git a/src/lix_logger.hpp b/src/lix_logger.hpp
index 4a34544..ae7751b 100644
--- a/src/lix_logger.hpp
+++ b/src/lix_logger.hpp
@@ -15,15 +15,15 @@ namespace xeus_lix
         explicit JupyterLogger(interpreter* interp);
 
         // redirects general log messages
-        void log(nix::Verbosity lvl, std::string_view s) override;
+        nix::Logger::BufferState log(nix::Verbosity lvl, std::string_view s) override;
         // redirects structured error info
-        void logEI(const nix::ErrorInfo& ei) override;
+        nix::Logger::BufferState logEI(const nix::ErrorInfo& ei) override;
 
         // the following are part of the nix::Logger interface but not implemented yet
-        void startActivity(nix::ActivityId, nix::Verbosity, nix::ActivityType, const std::string&, const Fields&, nix::ActivityId)
+        nix::Logger::BufferState startActivityImpl(nix::ActivityId, nix::Verbosity, nix::ActivityType, const std::string&, const Fields&, nix::ActivityId)
             override;
-        void stopActivity(nix::ActivityId) override;
-        void result(nix::ActivityId, nix::ResultType, const Fields&) override;
+        nix::Logger::BufferState stopActivityImpl(nix::ActivityId) override;
+        nix::Logger::BufferState resultImpl(nix::ActivityId, nix::ResultType, const Fields&) override;
 
     private:
         interpreter* p_interpreter;
diff --git a/src/lix_repl_commands.cpp b/src/lix_repl_commands.cpp
index 3f3ee57..99a99c0 100644
--- a/src/lix_repl_commands.cpp
+++ b/src/lix_repl_commands.cpp
@@ -98,12 +98,11 @@ namespace xeus_lix
     void interpreter::repl_load(const std::string& arg)
     {
         auto path = m_aio->blockOn(nix::lookupFileArg(*m_evaluator, arg)).unwrap();
-        nix::Value v(nix::Value::null_t{}), v_autocalled(nix::Value::null_t{});
-        m_evalState->evalFile(path, v);
+        nix::Value v = m_evalState->evalFile(path);
         nix::Bindings* auto_args = m_evaluator->buildBindings(0).finish();
-        m_evalState->autoCallFunction(*auto_args, v, v_autocalled, nix::noPos);
+        nix::Value v_autocalled = m_evalState->autoCallFunction(*auto_args, v, nix::noPos);
         m_evalState->forceAttrs(v_autocalled, nix::noPos, "while loading file attributes");
-        add_to_scope(*v_autocalled.attrs);
+        add_to_scope(*v_autocalled.attrs());
         m_loaded_files.push_back(path.to_string());
     }
 
@@ -149,7 +148,7 @@ namespace xeus_lix
         nix::Value v(nix::Value::null_t{});
         eval_pure_expression(arg, v);
         m_evalState->forceAttrs(v, nix::noPos, "while evaluating attribute set for :add");
-        add_to_scope(*v.attrs);
+        add_to_scope(*v.attrs());
     }
 
     // :b <expr> - Build a derivation
@@ -293,20 +292,22 @@ namespace xeus_lix
             return;
         }
         auto flakeRef = nix::parseFlakeRef(std::string(arg), nix::absPath("."), true);
-        nix::Value v(nix::Value::null_t{});
-        nix::flake::callFlake(
+        nix::Value v = nix::flake::callFlake(
             *m_evalState,
             nix::flake::lockFlake(
                 *m_evalState,
                 flakeRef,
                 nix::flake::LockFlags{ .updateLockFile = false,
                                        .useRegistries = !nix::evalSettings.pureEval.get(),
-                                       .allowUnlocked = !nix::evalSettings.pureEval.get() }
-            ),
-            v
+                                       .allowUnlocked = !nix::evalSettings.pureEval.get(),
+                                       .referenceLockFilePath = {},
+                                       .outputLockFilePath = {},
+                                       .inputOverrides = {},
+                                       .inputUpdates = {} }
+            )
         );
         m_evalState->forceAttrs(v, nix::noPos, "while loading flake outputs");
-        add_to_scope(*v.attrs);
+        add_to_scope(*v.attrs());
     }
 
     // :p <expr> - Evaluate and print expression recursively Strings are printed directly, without escaping.
@@ -328,7 +329,6 @@ namespace xeus_lix
                 v,
                 nix::PrintOptions{ .ansiColors = true,
                                    .force = true,
-                                   .derivationPaths = true,
                                    .maxDepth = std::numeric_limits<unsigned int>::max(),
                                    .prettyIndent = 2 }
             );
